### ISSUE TITLE
Chunk load/unload thrash due to entities ticking on the edges of ticketed chunks

### ISSUE THREAD

**Issue by Ricket** [MinecraftForge/MinecraftForge#5406]
Forge version `14.23.5.2781`
Pack `Enigmatica 2 Expert 1.51` (plus `ChickenChunks-1.12.2-2.4.1.73-universal.jar`)

My server has been having performance issues. What it seems to be is the following:

I have a ChickenChunks chunk loader loading some area. This is the picture from ChickenChunks' `/chunkloaders` command, green squares are chunks loaded due to the loader, red squares are loaded due to mod interactions.

![image](https://user-images.githubusercontent.com/196978/51807516-57a88980-224d-11e9-93ee-b1fdb866c423.png)

It quickly it turns to this:

![image](https://user-images.githubusercontent.com/196978/51807520-655e0f00-224d-11e9-81de-879e58985870.png)

But over time it goes back to the first picture, then the second, back and forth. The chunks are thrashing, loading and unloading, and every time there is noticeable lag as the server loads and unloads the chunk data.

I built and used this [chunkchecker mod](https://github.com/tornewuff/chunkchecker) to get stack traces and they all look like this:

```
[17:21:46] [Server thread/WARN] [chunkchecker]: Chunk at [0; 10, -18] has reloaded too many times, most recent trace: java.lang.Throwable
        at uk.org.wolfpuppy.minecraft.chunkchecker.ChunkChecker.chunkLoaded(ChunkChecker.java:104)
        at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_394_ChunkChecker_chunkLoaded_Load.invoke(.dynamic)
        at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90)
        at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:182)
        at net.minecraft.world.chunk.Chunk.func_76631_c(Chunk.java:862)
        at net.minecraft.world.gen.ChunkProviderServer.loadChunk(ChunkProviderServer.java:108)
        at net.minecraft.world.gen.ChunkProviderServer.func_186028_c(ChunkProviderServer.java:89)
        at net.minecraft.world.gen.ChunkProviderServer.func_186025_d(ChunkProviderServer.java:135)
        at net.minecraft.world.World.func_72964_e(World.java:310)
        at net.minecraft.world.ChunkCache.<init>(ChunkCache.java:37)
        at net.minecraft.pathfinding.PathNavigate.func_179680_a(SourceFile:113)
        at net.minecraft.pathfinding.PathNavigateGround.func_179680_a(SourceFile:65)
        at net.minecraft.pathfinding.PathNavigate.func_75488_a(SourceFile:93)
        at net.minecraft.pathfinding.PathNavigate.func_75492_a(SourceFile:143)
        at net.minecraft.entity.ai.EntityAIWander.func_75249_e(SourceFile:65)
        at net.minecraft.entity.ai.EntityAITasks.func_75774_a(SourceFile:102)
        at net.minecraft.entity.EntityLiving.func_70626_be(EntityLiving.java:763)
        at net.minecraft.entity.EntityLivingBase.func_70636_d(EntityLivingBase.java:2351)
        at net.minecraft.entity.EntityLiving.func_70636_d(EntityLiving.java:577)
        at net.minecraft.entity.monster.EntityMob.func_70636_d(EntityMob.java:45)
        at net.minecraft.entity.monster.EntityZombie.func_70636_d(EntityZombie.java:223)
        at net.minecraft.entity.EntityLivingBase.func_70071_h_(EntityLivingBase.java:2171)
        at net.minecraft.entity.EntityLiving.func_70071_h_(EntityLiving.java:295)
        at net.minecraft.entity.monster.EntityMob.func_70071_h_(EntityMob.java:50)
        at net.minecraft.world.World.func_72866_a(World.java:1993)
        at net.minecraft.world.WorldServer.func_72866_a(WorldServer.java:832)
        at net.minecraft.world.World.func_72870_g(World.java:1955)
        at net.minecraft.world.World.func_72939_s(World.java:1759)
        at net.minecraft.world.WorldServer.func_72939_s(WorldServer.java:613)
        at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:767)
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:397)
        at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:668)
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526)
        at java.lang.Thread.run(Thread.java:748)
```

So what I'm piecing together is that mobs near the edges of my loaded chunk area, are doing pathfinding which loads other chunks.

From reading this page: http://technical-minecraft.wikia.com/wiki/Chunk_Loading
It sounds to me like _normally_, there's a 2-chunk buffer zone around entities, and in that zone chunks are loaded but entities are not ticking, only tile entities and generation. That gives enough room so that entities in the center can pathfind without causing new chunk loads (unless they successfully _move_ into one of those chunks, at which time I guess they freeze & unload? or maybe they cause a few more chunks to load?).

But it seems that forge chunk tickets are fully loaded with entity ticking, and don't have that 2-chunk border around them, so any entities along the edges will cause new chunk loading when they pathfind.

That's my theory as to what's causing the lag, and if I'm correct, then one possible solution is that forge should additionally load a 2-chunk radius around any ticket chunks, those chunks being non-entity chunks.

I have tried expanding the chunk loading, but it only expands these pathfinding-loaded chunks.

I'm a bit stuck at this point, please let me know what else I should try or run in my server to diagnose this or fix/workaround it.

**Comment by Niko-sk2x** [comment]
You are definitely right about what is causing the issues but I really don't have any good idea for how it could be fixed. Last time I saw someone point out this issue, the reply I saw was more or less "the chunkloader mod should just deal with it". But there clearly is no right way to do it. 

Normally, when ticking entities vanilla is checking if chunks within radius of 2 chunks are loaded. This is the 2 chunks buffer radius. 

Forge patches this with a special case for forced chunks (what chunkloaded chunks are named in code), so that entities within forced chunks are always ticked, even with no chinks loaded around. And because of that, even adding extra forced chunks to have a buffer won't fix it because these extra chunks will also have their entities updated. 

This also may not be easily noticeable in a fresh world in dev environment because forced chunks don't have any entity spawning, so basically only animals (and maybe some mod-spawned entities) are kept in those chunks. 

One way to get around it would be to give mods control over whether a given forced chunk will get ticked or not. This would allow chunkloader mods to request extra non-ticked buffer chunks. Another way would be to remove the special case for forced chunks.

**Comment by williewillus** [comment]
so worth noting is the new vanilla `/forceload` command. I haven't looked at what they do yet, but we should see if they add extra buffer chunks to make the forced chunks entity processing, etc, and follow suit. The closer we are to vanilla behaviour the better, imo.
(I think Panda4994 did a video on this?)

**Comment by Niko-sk2x** [comment]
~~I don't think there is any chance of back porting any of the 1.14 chunk related changes to 1.12 or 1.13. The 1.14 chunk handling code is so weird I wasn't yet fully able to figure out how it works despite a few days of effort I put into it. I think it does add a buffer but it's done in such a weird way that I don't know for sure.~~

~~From what I've been able to tell so far, there is a class that handles some sort of levels shared between lighting propagation code and chunk tracking code. At first is looks like something designed for lighting code... But a bit backwards. Lighting levels passing through that class are actually inverted.~~

~~For chunk tracking, chunks also have some kind of level assigned to them and it appears to be updated on a way somewhat similar to how light propagates. The fact that at least 3 different kind of position as long coords are used all over that code also doesn't help understanding what it's actually doing.~~

Apparently this exists in 1.13.1, which makes all of the above irrelevant. I didn't notice it because forge 1.13 branch isn't using 1.13.1 yet.

**Comment by williewillus** [comment]
forceload is from 1.13.1 so hopefully it's relatively decipherable

**Comment by Ricket** [comment]
It sounds like there were a few ideas on how to fix/backport this, should we keep this issue open?

**Comment by Ricket** [comment]
This is still an issue, right? Especially in 1.12 (still the majority mod version, right?) it was never fixed afaik.

**Comment by pupnewfster** [comment]
1.12 is no longer supported https://www.minecraftforge.net/forum/topic/79304-112114-and-lts-system/ not sure if it is still an issue in 1.14/1.15

**Comment by Niko-sk2x** [comment]
As far as I know it has been completely fixed by mojang in 1.14 with new chunk management system.
